# pandas - Parsing Timestamps

In [1]:
import pandas as pd
import numpy as np

In [5]:
%matplotlib inline

chronological = theo thời gian

Timestamps are common in sensor data or other time series datasets.

In [7]:
tags=pd.read_csv("ml-20m/tags.csv",sep=",")
tags.dtypes

userId        int64
movieId       int64
tag          object
timestamp     int64
dtype: object

minh co the thay timestamp o day co dinh dang int64, nen gio minh se convert    

In [8]:
tags["parsed_time"]=pd.to_datetime(tags["timestamp"],unit='s')

In [12]:
tags["parsed_time"].dtypes

dtype('<M8[ns]')

u' u`, da chuyen dinh dang int64 sang datetime64[ns]

In [11]:
tags.head()

userId  movieId            tag   timestamp         parsed_time
0      18     4141    Mark Waters  1240597180 2009-04-24 18:19:40
1      65      208      dark hero  1368150078 2013-05-10 01:41:18
2      65      353      dark hero  1368150079 2013-05-10 01:41:19
3      65      521  noir thriller  1368149983 2013-05-10 01:39:43
4      65      592      dark hero  1368150078 2013-05-10 01:41:18

#kieu du lieu ban dau timestamp o day la, kieu interger

# selecting rows based on timestamps

In [15]:
greater_than_t=tags["parsed_time"]> '2015-02-01'
select_rows=tags[greater_than_t]
tags.shape,select_rows.shape

((465564, 5), (12130, 5))

o day co nghia la co 12130 hang co so ngay lon hon 2015-02-01

In [16]:
select_rows

userId  movieId                       tag   timestamp  \
301        318      260                     1970s  1424472169   
302        318      260                   fantasy  1424472169   
303        318      260                    sci-fi  1424472169   
304        318   115149                    Action  1424534310   
305        318   115149                   Revenge  1424534283   
306        318   115149              Willem Dafoe  1424534383   
324        342      318                masterplan  1423424898   
326        342      318             prison escape  1423424898   
327        342      318                redemption  1423424898   
589        342   108190           based on a book  1426978115   
590        342   108190                  dystopia  1426978076   
1018       370     6975  breaking the fourth wall  1425242234   
1019       370    27592                   violent  1425241975   
2203       969       97            class conflict  1426754226   
2204       969      288                  mindfuck  1426752631   
2205       969      288                  paranoid  1426752639   
2206       969      288               psychedelic  1426752634   
2207       969     1263       american propaganda  1426753995   
2208       969     3742            class conflict  1426753876   
2209       969     3742                   Classic  1426754311   
2210       969     3742                Propaganda  1426754337   
2211       969     4848               Atmospheric  1426752575   
2212       969     4848                      dark  1426752581   
2213       969     4848                  Mindfuck  1426752569   
2214       969     6530            claustrophobic  1426753144   
2215       969     6530            mental illness  1426751285   
2216       969     6530                   suicide  1426751289   
2217       969     8950                depressing  1426749117   
2218       969     8950               psychodrama  1426752695   
2219       969     8950             schizophrenia  1426752689   
...        ...      ...                       ...         ...   
461895  137227   120222         horribly animated  1422821079   
461896  137227   120222              nazi imagery  1422821051   
461897  137227   120222                    racist  1422821047   
461898  137227   120222        racist stereotypes  1422821071   
462190  137249      296               influential  1423437040   
462191  137249      296                  neo-noir  1423437040   
462192  137249      296         quentin tarantino  1423437040   
462195  137277   110052     marcello mastroianni,  1422925214   
462979  137924    46855                 Criterion  1427722627   
462980  137924    46855         French Resistance  1427722622   
462981  137924    46855              World War II  1427722614   
463004  138063     4355                    Hockey  1424230535   
463005  138063   108160                    Hockey  1424230501   
463504  138103    67894                   finnish  1423255773   
463509  138103    69442                   finnish  1423255750   
463544  138103    73501                   finnish  1423255756   
463545  138103    74681                   finnish  1423255760   
463573  138103    84553                   finnish  1423255764   
464786  138301    52042                 espionage  1426032479   
464787  138301    52042       resistance movement  1426032481   
464788  138301    52042              World War II  1426032485   
464958  138301   109487                 ambitious  1423178287   
464959  138301   109487                atmosphere  1423178302   
464960  138301   109487                plot holes  1423178186   
464961  138301   109487                    sci-fi  1423178227   
464962  138301   109487           science fiction  1423178255   
464963  138301   109487            Self-Indulgent  1423178219   
464964  138301   109487                     space  1423178343   
464965  138301   109487        visually appealing  1423178319   
464966  138301   109487          wasted pot

# sorting the table using the timestamps

In [20]:
tags.sort_values(by="parsed_time",ascending=False)[:10]

userId  movieId                              tag   timestamp  \
339178  102853   115149                    russian mafia  1427771352   
158780   46072     6058                      premonition  1427760764   
158763   46072     3409                      premonition  1427760726   
288375   87797      215                           Vienna  1427755801   
290535   88044   106782                        profanity  1427754096   
290527   88044   106782                            crime  1427753921   
290526   88044   106782                           Amoral  1427753913   
290531   88044   106782                       Jonah Hill  1427753849   
290530   88044   106782                       inaccurate  1427753806   
290528   88044   106782  Economically Illiterate Writers  1427753739   

               parsed_time  
339178 2015-03-31 03:09:12  
158780 2015-03-31 00:12:44  
158763 2015-03-31 00:12:06  
288375 2015-03-30 22:50:01  
290535 2015-03-30 22:21:36  
290527 2015-03-30 22:18:41  
290526 2015-03-30 22:18:33  
290531 2015-03-30 22:17:29  
290530 2015-03-30 22:16:46  
290528 2015-03-30 22:15:39

sap xep xem thang nao gan day nhat(moi nhat)

# Average movie ratings over Time

Are movie ratings related to the year of launch?

In [23]:
ratings=pd.read_csv("ml-20m/ratings.csv",sep=",")
average_rating=ratings[['movieId','rating']].groupby('movieId',as_index=False)

In [24]:
average_rating.tail()

movieId  rating
8843        27359     2.5
21166        3332     3.0
27111       91213     1.0
27152      107881     2.5
27154      109492     3.5
29578      100946     4.5
31739       46957     3.5
31814       83417     3.5
31863      100062     3.5
33559      107230     5.0
46231       68337     4.0
46326       83291     4.0
60901       98217     2.5
62737       77355     1.0
65337       38839     3.0
65416       80330     3.0
65472       91981     3.5
65510       99904     1.0
65517      103361     3.5
65518      104064     3.5
65519      104066     3.0
65520      104464     4.0
65521      104638     1.5
65522      105753     3.5
69410       88061     3.0
72715       43248     1.0
72739       87591     4.0
72744       93816     4.0
79282       40679     2.0
79318       44892     1.5
...           ...     ...
20000233    50872     3.5
20000234    51086     3.5
20000235    51662     4.5
20000236    51884     4.5
20000237    52579     4.0
20000238    52975     4.0
20000239    53123     4.0
20000240    53125     3.0
20000241    53322     4.0
20000242    53464     4.0
20000243    53996     4.5
20000244    55269     5.0
20000245    55814     5.0
20000246    56757     3.0
20000247    56801     3.0
20000248    58879     4.5
20000249    59315     4.0
20000250    59725     3.0
20000251    59784     5.0
20000252    60069     4.0
20000253    60816     4.5
20000254    61160     4.0
20000255    65682     4.5
20000256    66762     4.5
20000257    68319     4.5
20000258    68954     4.5
20000259    69526     4.5
20000260    69644     3.0
20000261    70286     5.0
20000262    71619     2.5

[107964 rows x 2 columns]

In [26]:
movies=pd.read_csv("ml-20m/movies.csv",sep=",")

In [27]:
joined=movies.merge(average_rating, on='movieId',how='inner')
joined.head()
joined.corr()

ValueError: can not merge DataFrame with instance of type <class 'pandas.core.groupby.DataFrameGroupBy'>